cpp/src/fil/fil.cu

```diff
@@ -81,6 +81,9 @@ struct forest {
     int max_shm = 0;
     CUDA_CHECK(cudaDeviceGetAttribute(
       &max_shm, cudaDevAttrMaxSharedMemoryPerBlockOptin, device));
+    ASSERT(max_shm < 262144,
```

Comment:
Our GPUs have been growing the shared memory size generation after generation. Eventually, a CUDA GPU might come by that supports more shared memory that would fit into unsigned 16-bit int. For such a GPU, we would have otherwise silently overflowed the index calculation due to short division. It would have failed cpp tests, but we might forget about this source of bugs, if not for the failing assert.

Reply:
It would be better if this explanation were a comment in the code.